infrastructure/zk/src/database.ts

```diff
@@ -1,101 +1,233 @@
+}
+
+async function resetTestDal(dalPath: DalPath, dbUrl: string) {
     console.log('recreating postgres container for unit tests');
     await utils.spawn('docker compose -f docker-compose-unit-tests.yml down');
     await utils.spawn('docker compose -f docker-compose-unit-tests.yml up -d');
     await wait(100);
     console.log('setting up a database template');
-    await setup();
+    await setupForDal(dalPath, dbUrl);
     console.log('disallowing connections to the template');
     await utils.spawn(
-        `psql "${process.env.DATABASE_URL}" -c "update pg_database set datallowconn = false where datname = current_database()"`
+        `psql "${dbUrl}" -c "update pg_database set datallowconn = false where datname = current_database()"`
     );
 }
 
-export async function drop() {
+export async function resetTest(opts: any) {
     await utils.confirmAction();
-    console.log('Dropping DB...');
-    await utils.spawn('cargo sqlx database drop -y');
+    let dals = getTestDals(opts);
+    for (const [dalPath, dbUrl] of dals.entries()) {
+        await resetTestDal(dalPath, dbUrl);
+    }
 }
 
-export async function migrate() {
+async function dropForDal(dalPath: DalPath, dbUrl: string) {
+    console.log(`Dropping DB for dal ${dalPath}...`);
+    await utils.spawn(`cargo sqlx database drop -y --database-url ${dbUrl}`);
+}
+
+export async function drop(opts: any) {
     await utils.confirmAction();
-    console.log('Running migrations...');
-    await utils.spawn('cd core/lib/dal && cargo sqlx database create && cargo sqlx migrate run');
+    let dals = getDals(opts);
+    for (const [dalPath, dbUrl] of dals.entries()) {
+        await dropForDal(dalPath, dbUrl);
+    }
+}
+
+async function migrateForDal(dalPath: DalPath, dbUrl: string) {
+    console.log(`Running migrations for ${dalPath}...`);
+    await utils.spawn(
+        `cd ${dalPath} && cargo sqlx database create --database-url ${dbUrl} && cargo sqlx migrate run --database-url ${dbUrl}`
+    );
+}
+
+export async function migrate(opts: any) {
+    await utils.confirmAction();
+    let dals = getDals(opts);
+    for (const [dalPath, dbUrl] of dals.entries()) {
+        await migrateForDal(dalPath, dbUrl);
+    }
+}
+
+async function generateMigrationForDal(dalPath: DalPath, dbUrl: string, name: String) {
+    console.log(`Generating migration for ${dalPath}...`);
+    await utils.spawn(`cd ${dalPath} && cargo sqlx migrate add -r ${name}`);
 }
 
-export async function generateMigration(name: String) {
+export async function generateMigration(opts: any) {
+    if ((!opts.prover && !opts.server) || (opts.prover && opts.server)) {
+        throw new Error(
+            '[aborted] please specify a single database to generate migration for (i.e. to generate a migration for server `zk db new-migration --server name_of_migration`'
+        );
+    }
+    let name = '';
+    if (opts.prover) {
+        name = opts.prover;
+    }
+    if (opts.server) {
+        name = opts.server;
+    }
```

Comment:
This is a bit clunky. Why not just introduce a public enum `export enum DBType { Server, Prover }` and accept it as an argument?

Reply:
I think TypeScript should be able to coerce interfaces, e.g. if you declare your own interface that matches the interface of what Commander would give you, it should work.

I've checked out your branch, and the following seems to compile just fine:

```
function getDals(opts: { server: boolean, prover: boolean }): Map<DalPath, string>
```

as well as 

```
export interface Opts {
    server: boolean,
    prover: boolean,
}

function getDals(opts: Opts): Map<DalPath, string> {
```